src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs

```diff
@@ -0,0 +1,459 @@
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!_disposed)
+            {
+                if (disposing)
+                {
+#if NETSTANDARD1_4
+                    if (_rsa != null && _disposeRsa)
+                        _rsa.Dispose();
+#else
+                    if (_rsaCryptoServiceProvider != null && _disposeRsa)
+                        _rsaCryptoServiceProvider.Dispose();
+#endif
+
+                    if (_rsaCryptoServiceProviderProxy != null)
+                        _rsaCryptoServiceProviderProxy.Dispose();
+
+                    _disposed = true;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Checks if an algorithm is supported.
+        /// </summary>
+        /// <param name="key">The <see cref="SecurityKey"/> that will be used for crypto operations.</param>
+        /// <param name="algorithm">The KeyWrap algorithm to apply.</param>
+        /// <returns>true if the algorithm is supported; otherwise, false.</returns>
+        protected virtual bool IsSupportedAlgorithm(SecurityKey key, string algorithm)
+        {
+            if (key == null)
+                return false;
+
+            if (string.IsNullOrEmpty(algorithm))
+                return false;
+
+            if (algorithm.Equals(SecurityAlgorithms.RsaPKCS1, StringComparison.Ordinal)
+                || algorithm.Equals(SecurityAlgorithms.RsaOAEP, StringComparison.Ordinal)
+                || algorithm.Equals(SecurityAlgorithms.RsaOAEP256, StringComparison.Ordinal))
+            {
+                if (key as RsaSecurityKey != null)
+                    return true;
+
+                X509SecurityKey x509Key = key as X509SecurityKey;
+                if (x509Key != null)
+                {
+#if NETSTANDARD1_4
+                    if (x509Key.PublicKey as RSA != null)
+                        return true;
+#else
+                    if (x509Key.PublicKey as RSACryptoServiceProvider != null)
+                        return true;
+#endif
+
+                    return false;
+                }
+
+                var jsonWebKey = key as JsonWebKey;
+                if (jsonWebKey.Kty == JsonWebAlgorithmsKeyTypes.RSA)
+                    return true;
+            }
+
+            return false;
+        }
+
+#if NETSTANDARD1_4
+        /// <summary>
+        /// Initialize RSA algorithm.
+        /// </summary>
+        /// <param name="key">The <see cref="SecurityKey"/> that will be used for crypto operations.</param>
+        /// <param name="algorithm">The RSA KeyWrap algorithm to apply.</param>
+        /// <param name="willDecrypt">Whether this <see cref="RsaKeyWrapProvider"/> is required to create decrypts then set this to true.</param>
+        /// <exception cref="ArgumentOutOfRangeException">The <see cref="SecurityKey"/> is not supported.</exception>
+        protected virtual void ResolveRsaAlgorithm(SecurityKey key, string algorithm, bool willDecrypt)
```

Comment:
This is very similar to: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/master/src/Microsoft.IdentityModel.Tokens/AsymmetricSignatureProvider.cs#L319
Can we have a single function?